File: app/src/main/java/com/liadpaz/amp/viewmodels/Artist.java
package com.liadpaz.amp.viewmodels;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import java.util.ArrayList;

public class Artist implements Parcelable {
    public static final Creator<Artist> CREATOR = new Creator<Artist>() {
        @NonNull
        @Override
        public Artist createFromParcel(@NonNull Parcel in) {
            return new Artist(in);
        }

        @NonNull
        @Override
        public Artist[] newArray(int size) {
            return new Artist[size];
        }
    };

    public final String name;
    public final ArrayList<Song> songs;

    public Artist(@NonNull String name, @NonNull ArrayList<Song> songs) {
        this.name = name;
        this.songs = songs;
    }

    private Artist(@NonNull Parcel in) {
        name = in.readString();
        songs = in.createTypedArrayList(Song.CREATOR);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeTypedList(songs);
    }
}

File: app/src/main/java/com/liadpaz/amp/viewmodels/Artist.kt
package com.liadpaz.amp.viewmodels

import android.os.Parcel
import android.os.Parcelable
import com.liadpaz.amp.viewmodels.Song
import java.util.*
import kotlin.collections.ArrayList

data class Artist(val name: String, val  songs: ArrayList<Song>) : Parcelable {
    private constructor(parcel: Parcel) : this(parcel.readString()!!, parcel.createTypedArrayList(Song.CREATOR)!!)

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(name)
        dest.writeTypedList(songs)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (this.javaClass != other?.javaClass) return false
        other as Artist
        return other.name == name && other.songs == songs
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + songs.hashCode()
        return result
    }

    companion object CREATOR : Parcelable.Creator<Artist> {
        override fun createFromParcel(parcel: Parcel): Artist {
            return Artist(parcel)
        }

        override fun newArray(size: Int): Array<Artist?> {
            return arrayOfNulls(size)
        }
    }
}